sklearn/utils/_array_api.py

```diff
@@ -822,6 +822,84 @@ def _estimator_with_converted_arrays(estimator, converter):
     return new_estimator
 
 
+def make_converter(X):
+    """Helper to implement the 'y follows X' rule.
+
+    Returns a function that converts an array to the namespace and device of X.
+
+    When X is not an array api array, the converter does nothing.
+    """
+    if isinstance(X, numpy.ndarray):
+
+        def convert(data):
+            if not hasattr(data, "__dlpack__"):
+                return data
+            data_xp, _ = get_namespace(data)
+            return _convert_to_numpy(data, data_xp)
+
+        return convert
+    xp, is_array_api, device_ = get_namespace_and_device(X)
+    if not is_array_api:
+
+        def convert(data):
+            return data
+
+        return convert
+
+    def convert(data):
+        if not isinstance(data, numpy.ndarray) and not hasattr(data, "__dlpack__"):
+            return data
+        data_xp, _, data_device = get_namespace_and_device(data)
+        if data_xp == xp and data_device == device_:
+            return data
+        try:
+            return xp.asarray(data, device=device_)
+        except Exception:  # pragma: no cover
+            # direct conversion to a different library may fail in which
+            # case we try converting to numpy first
+            data = _convert_to_numpy(data, data_xp)
+            return xp.asarray(data, device=device_)
+
+    return convert
+
+
+def convert_attributes(estimator, ref_array):
+    """
+    Convert attributes of estimator to namespace and device of reference array.
+
+    Attributes which are not arrays are left unchanged.
+    """
+    return _estimator_with_converted_arrays(estimator, make_converter(ref_array))
```

Comment:
Could you explain why we need `make_converter`? Could we have the same thing by having a function like `convert_to(reference, X)` which is used here a bit like ` _estimator_with_converted_arrays(estimator, functools.partial(convert_to, ref_array))`?

I'm not sure which approach I prefer, so just trying to understand it at the moment.

Reply:
>  Could we have the same thing by having a function like convert_to(reference, X) which is used here a bit like  _estimator_with_converted_arrays(estimator, functools.partial(convert_to, ref_array))?

yes, that would work just as well! I had `make_converter` return a function because I wanted to use the existing `_estimator_with_converted_arrays`, which expects a function. But we can also define `convert_to` and use `partial` as you suggest, I don't really have a preference

Reply:
(note both options will not have the same behavior when the `dispatch_array_api` config changes between the call to `make_converter/partial` and using the result, but this should probably be avoided in any case)